### ISSUE TITLE
Please help advise on multi-gpu training of marblenet with pytorch lightning 1.9.4

### ISSUE THREAD

**Issue by PaulSZH95** [NVIDIA/NeMo#6961]
### Discussed in https://github.com/NVIDIA/NeMo/discussions/6959

<div type='discussions-op-text'>

<sup>Originally posted by **PaulSZH95** July  2, 2023</sup>
Hi all, 
1) May i ask if it is normal for single gpu to have lesser % utilisation than if ddp was used.
2) May I ask if it is normal that if i have 2 gpu and 1 gpu is at 100% the other would be at only 30 %
3) how do i track what is normal gpu utilisation with multigpu training
4) are these methods : nemo.collections.asr.modules.AudioToMFCCPreprocessor and nemo.collections.asr.modules.SpectrogramAugmentation using cpu or gpu
5) if the methods uses cpu, how will if affect parallel training. Details below

I followed the tutorial on VAD through https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/starthere/tutorials.html which used branch r1.18.0. I have set up all dependencies i.e. having megatron and apex. My pytorch compatibility are also resolved.

According to the tutorial I can speed up training with device = 2 if I have 2 device.

While the image I provide is a snapshot where both gpu has the same load, most of the time it would be either 1 gpu on 80 - 100% and the other at 0% - 30 % or the other way around. So both gpu are used but at any 1 time the load on either gpu would be vastly different. I am not sure if it is an expected behavior.

In addition, when i use the default marble_net config as given in tutorial, the gpu utilisation is at 14 - 20 % which is much diff than if ddp was used.

Using watch -n 0.5 nvidia-smi (please ignore the circled area)
![image - Copy](https://github.com/NVIDIA/NeMo/assets/85733116/850a0bd6-5f7c-4584-a33c-1e0a4fa82f84)

Lastly, unlike other discussion where they found the pid to be allocated to just 1 gpu, it seems both of my gpu is running. Unfortunately I am still trying to get pid for gpu to be shown thus I can't say for sure either.
</div>

<div>
Could I also check how the number of batch is decided?

in the MarbleNet_3x2x64.yaml file, batch_size was clearly stated as 128. in log,

[NeMo I 2023-07-02 03:08:18 collections:298] Filtered duration for loading collection is 0.000000.
[NeMo I 2023-07-02 03:08:18 collections:301] Dataset loaded with 2164876 items, total duration of 378.85 hours.
[NeMo I 2023-07-02 03:08:18 collections:303] # 2164876 files loaded accounting to # 2 labels
[NeMo I 2023-07-02 03:08:22 collections:298] Filtered duration for loading collection is 0.000000.
[NeMo I 2023-07-02 03:08:22 collections:301] Dataset loaded with 234914 items, total duration of 41.11 hours.
[NeMo I 2023-07-02 03:08:22 collections:303] # 234914 files loaded accounting to # 2 labels
[NeMo I 2023-07-02 03:08:34 collections:298] Filtered duration for loading collection is 0.000000.
[NeMo I 2023-07-02 03:08:34 collections:301] Dataset loaded with 824574 items, total duration of 144.30 hours.

but in training:

Epoch 3: 100%|████████████████████████████████████████████████████| 18750/18750 [59:59<00:00, 5.21it/s, loss=0.322, v_num=8-42

if 18750 at batch_size of 128 I would have to have 2400000 which is severely greater than my 2164876 training items. And even if the batch size is the measurement of mfcc, each file give 64 x 64, then number of batch should be at 1082438. if purely by audio file then it should be 16913 or 16914 depending on rounding.

please someone help, thank you.

</div>

**Comment by titu1994** [comment]
1) no it's not normal for 1 GPU to have a large difference with and without DDP. 
2) no it's not usually to have such large GPU utilization difference. Use killall -9 python and then run code to make sure there's no dead cuda process. 
3) Nvidia smi is easiest way
4) both use gpu. 

Apex is not needed for ASR domain. 

The GPU load may change over time, the beginning may use more utilization for setting up dataloaders. But it should not continue to be like that and stabilize to similar values between GPUs. However in the picture they are 20% - that's fine. You can increase batch size since you have large memory on your GPU to increase GPU utilization. 

The progress bar includes the validation and test set samples. That's why your number of batch steps is more than just trian set

**Comment by PaulSZH95** [comment]
> Apex is not needed for ASR domain.

Thanks for the reply and thanks for highlighting this.  I had encountered the warning of could not import distributed_fused_adam optimizer from Apex when I downloaded only asr component of nemo and thought it was a necessary component.

**Comment by PaulSZH95** [comment]
Hi sir @titu1994,

may I trouble you again.

On running ddp protocol with 2 gpu, one of my gpu is constantly hitting near 100 % and the other at less than 20 %. Is this normal? I have tried to  increase my num_workers to 16 (max on my system) and also down to 0. In both instances 1 gpu_util is still significantly more than the others.

![image](https://github.com/NVIDIA/NeMo/assets/85733116/2f2b7924-eda5-42b0-9cd0-43bfc3c4ef93)

lots of thanks.

**Comment by PaulSZH95** [comment]
Hi Mr @titu1994 ,

May I have your help again. I have encountered the issue below and had experimented with num_workers from 0 to 16 (max my system could do) to no avail.

I am running training on docker instances therefore there should be no dead python processes as far as I could tell. 

The difference in gpu-util between my gpu-0 and gpu-1 is extremely wide.

![image](https://github.com/NVIDIA/NeMo/assets/85733116/383f47bf-5a40-4877-95ac-98443f0b0dc7)

There's also no dead or ghost python processes:

![image](https://github.com/NVIDIA/NeMo/assets/85733116/c640228c-933c-44cf-8edf-7ac5e5ef8309)

**Comment by titu1994** [comment]
@fayejf was there any changes to code for speech classification models ?

**Comment by titu1994** [comment]
We haven't updated that code in quite some time so I don't think the issue is occuring due to some bad change. 100% on gpu usually means some sort of hang

**Comment by PaulSZH95** [comment]
@titu1994 Hi, could I please have your help again. how I could get reproducible results? I have seeded my config file as follows:

and used the following settings individually and together but the train and validation loss per epochs still varies significantly.

deterministic= True
and
seed_everything()

    seed_everything(seed=42,workers=True)
    trainer = pl.Trainer(**config.trainer,deterministic=True)
    # trainer = pl.Trainer(**config.trainer)
    exp_dir = exp_manager(trainer, config.get("exp_manager", None))
    exp_dir = str(exp_dir)
    vad_model = nemo_asr.models.EncDecClassificationModel(
        cfg=config.model, trainer=trainer
    )
    trainer.fit(vad_model)


name: &name "MarbleNet-3x2x64"

model:
  sample_rate: 16000
  repeat: 2
  dropout: 0.0
  kernel_size_factor: 1.0

  labels: ['background', 'speech']

  train_ds:
    manifest_filepath: ???
    sample_rate: ${model.sample_rate}
    labels: ${model.labels}
    batch_size: 320
    shuffle: True
    # tarred datasets
    is_tarred: false
    tarred_audio_filepaths: null
    tarred_shard_strategy: "scatter"
    shuffle_n: 2048
    num_workers: 6
    pin_memory: True
    # bucketing params
    bucketing_strategy: "synced_randomized"
    bucketing_batch_size: null
    bucketing_weights: null
    seed: 42
    augmentor:
      shift:
        prob: 1.0
        min_shift_ms: -5.0
        max_shift_ms: 5.0
      white_noise:
        prob: 1.0
        min_level: -90
        max_level: -46

  validation_ds:
    manifest_filepath: ???
    sample_rate: ${model.sample_rate}
    labels: ${model.labels}
    batch_size: 320
    shuffle: False
    num_workers: 6
    pin_memory: true
    val_loss_idx: 0
    seed: 42

  test_ds:
    manifest_filepath: ???
    sample_rate: ${model.sample_rate}
    labels: ${model.labels}
    batch_size: 320
    shuffle: False
    num_workers: 6
    pin_memory: true
    test_loss_idx: 0
    seed: 42

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMFCCPreprocessor
    window_size: 0.025
    window_stride: 0.01
    window: "hann"
    n_mels: &n_mels 64
    n_mfcc: *n_mels
    n_fft: 512

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2
    time_masks: 2
    freq_width: 15
    time_width: 25
    rect_masks: 5
    rect_time: 25
    rect_freq: 15

  encoder:
    _target_: nemo.collections.asr.modules.ConvASREncoder
    feat_in: *n_mels
    activation: relu
    conv_mask: true

    jasper:
      - filters: 128
        repeat: 1
        kernel: [11]
        stride: [1]
        dilation: [1]
        dropout: ${model.dropout}
        residual: false
        separable: true
        kernel_size_factor: ${model.kernel_size_factor}

      - filters: 64
        repeat: ${model.repeat}
        kernel: [13]
        stride: [1]
        dilation: [1]
        dropout: ${model.dropout}
        residual: true
        separable: true
        kernel_size_factor: ${model.kernel_size_factor}

      - filters: 64
        repeat: ${model.repeat}
        kernel: [15]
        stride: [1]
        dilation: [1]
        dropout: ${model.dropout}
        residual: true
        separable: true
        kernel_size_factor: ${model.kernel_size_factor}

      - filters: 64
        repeat: ${model.repeat}
        kernel: [17]
        stride: [1]
        dilation: [1]
        dropout: ${model.dropout}
        residual: true
        separable: true
        kernel_size_factor: ${model.kernel_size_factor}

      - filters: 128
        repeat: 1
        kernel: [29]
        stride: [1]
        dilation: [2]
        dropout: ${model.dropout}
        residual: false
        separable: true
        kernel_size_factor: ${model.kernel_size_factor}

      - filters: &enc_final_filters 128
        repeat: 1
        kernel: [1]
        stride: [1]
        dilation: [1]
        dropout: ${model.dropout}
        residual: false

  decoder:
    _target_: nemo.collections.asr.modules.ConvASRDecoderClassification
    feat_in: *enc_final_filters
    return_logits: true
    pooling_type: 'avg'

  optim:
    name: sgd
    lr: 0.01
    # optimizer arguments
    weight_decay: 0.001
    momentum: 0.9

    # scheduler setup
    sched:
      name: PolynomialHoldDecayAnnealing
      # Scheduler params
      power: 2.0
      warmup_ratio: 0.05
      hold_ratio: 0.45
      min_lr: 0.001
      last_epoch: -1

trainer:
  devices: 1 # number of gpus
  max_epochs: 30
  max_steps: -1 # computed at runtime if not set
  num_nodes: 1
  accelerator: gpu
  # accelerator: cpu
  strategy: ddp
  accumulate_grad_batches: 1
  enable_checkpointing: False  # Provided by exp_manager
  logger: False  # Provided by exp_manager
  log_every_n_steps: 1  # Interval of logging.
  val_check_interval: 1.0  # Set to 0.25 to check 4 times per epoch, or an int for number of iterations
  benchmark: false # needs to be false for models with variable-length speech input as it slows down training

exp_manager:
  exp_dir: null
  name: *name
  create_tensorboard_logger: True
  create_checkpoint_callback: True
  create_wandb_logger: False
  wandb_logger_kwargs:
    name: null
    project: null
  early_stopping_callback_params:
    monitor: val_loss
    mode: min
    min_delta: 0.001
    patience: 5

**Comment by PaulSZH95** [comment]
noted with thanks
